feat(playground): Implement codemirror based template string editor

cephalization opened the pull request:
TODO:

- [x] Implement tests
  - [x] Need to figure out rollup plugins in jest or switch to vitest
- [x] Implement mustache like template language
  - [x] Should be a quick-ish port just copy-pasting the fstring grammar and adding double brackets
- [x] Implement variable replacement util that utilizes parsed tree

cephalization commented:
This is the last known mustache syntax bug I can find.

Triple backticks selects the left two most braces and the inner most two right braces and includes one brace as input. Desired behavior is probably to escape this entirely.

<img width="621" alt="image" src="https://github.com/user-attachments/assets/ef49218b-30d9-41ea-bf7d-57dc4a7a8da6">

cephalization commented on app/src/pages/playground/PlaygroundChatTemplate.tsx:
```diff
@@ -245,27 +241,12 @@ function SortableMessageItem({
-            variant="quiet"
-            aria-label={"Message content"}
+            aria-label="Message content"
+            templateLanguage={TemplateLanguages.Mustache}
```
Change this to `TemplateLanguages.FString` to test that instead.

Parker-Stafford commented on app/src/components/templateEditor/language/fStringTemplating/fStringTemplating.syntax.grammar:
```diff
@@ -0,0 +1,25 @@
+}
+
+@local tokens {
+  RBrace { "}" }
```
not sure if you can write comments in here but might be nice to have some since this is pretty foreign. 

Why is Rbrace in @local tokens and LBrace in @tokens?

cephalization replied:
This will be difficult information to capture within comments im afraid to say 😂 

I can link to relevant sections in the lezer docs if that helps

Parker-Stafford replied:
yeah maybe just linking to some docs at the top of these files would be nice!

cephalization replied:
Done

Parker-Stafford commented on app/src/components/templateEditor/language/fStringTemplating/fStringTemplating.ts:
```diff
@@ -0,0 +1,63 @@
+  languageData: {},
+});
+
+export const debugParser = (text: string) => {
+  const tree = FStringTemplatingLanguage.parser.parse(text);
+  return tree.toString();
+};
```
is this still needed? if needed for tests maybe just add a comment

Parker-Stafford commented on app/src/components/templateEditor/language/languageUtils.ts:
```diff
@@ -0,0 +1,86 @@
+  const tree = parser.parse(text);
+  const variables: string[] = [];
+  const cur = tree.cursor();
+  do {
+    if (cur.name === "Variable") {
+      variables.push(text.slice(cur.node.from, cur.node.to));
+    }
+  } while (cur.next());
```
any reason for a do ... while loop here? no big deal just not used very often so a bit harder to read

cephalization replied:
It was recommended by the lezer docs, I can drop a link. It's because the tree cursor starts at the first node and we only want to iterate after the first loop.

Parker-Stafford commented on app/src/components/templateEditor/language/languageUtils.ts:
```diff
@@ -0,0 +1,86 @@
+ *
+ * @returns An array of variable names.
+ */
+export const extractVariables = (parser: LRParser, text: string) => {
```
nit: we typically prefer "named" (object) arguments I would say for basically anything above 1 param

Parker-Stafford commented on app/src/components/templateEditor/language/languageUtils.ts:
```diff
@@ -0,0 +1,86 @@
+  /**
+   * The parser for the templating language.
+   *
+   * Should be MustacheLikeTemplatingLanguage or FStringTemplatingLanguage.
```
is there anyway to restrict this type? typeof MustacheLIke... etc.

cephalization replied:
Not really, the output parser type is too complex, and technically this function works with any LRParser.

It just won't find variables correctly unless you use one of the two langs we created or any other lang that outputs Variable nodes

Parker-Stafford commented on app/src/components/templateEditor/language/languageUtils.ts:
```diff
@@ -0,0 +1,86 @@
+   *
+   * If a variable is not found in this object, it will be left as is.
+   */
+  variables: Record<string, string | number | boolean>;
```
nit: your check below avoids this, but sometimes prefer writing records as Record<string, string | number | undefined> just because it's a little more honest and catches unset keys

cephalization replied:
in my opinion we should just enable `noUncheckedIndexedAccess` in our tsconfig if we are concerned about unset keys

cephalization replied:
Added

Parker-Stafford commented on app/src/components/templateEditor/language/mustacheLikeTemplating/mustacheLikeTemplating.syntax.grammar:
```diff
@@ -0,0 +1,25 @@
+@top MustacheLikeTemplate {(Template | Text | LEscape | REscape)*}
+
+
```
same here on some comments if possible

cephalization replied:
Done

Parker-Stafford commented on app/src/components/templateEditor/language/mustacheLikeTemplating/mustacheLikeTemplating.ts:
```diff
@@ -0,0 +1,59 @@
+
+import { parser } from "./mustacheLikeTemplating.syntax.grammar";
+
+// https://codemirror.net/examples/lang-package/
```
tiny thing can put this in the doc comment below with @see for easy navigation

Parker-Stafford commented on app/src/components/templateEditor/language/mustacheLikeTemplating/mustacheLikeTemplating.ts:
```diff
@@ -0,0 +1,59 @@
+  languageData: {},
+});
+
+export const debugParser = (text: string) => {
+  const tree = MustacheLikeTemplatingLanguage.parser.parse(text);
+  return tree.toString();
+};
```
same q here

Parker-Stafford reviewed (commented):
cool stuff

mikeldking commented:
I think there's some very minor syntax issues e.g. raw {{}}

<img width="457" alt="Screenshot 2024-10-11 at 5 59 36 PM" src="https://github.com/user-attachments/assets/2e7ad4bb-5f68-4ef2-8085-33247fb6f3b8">

mikeldking commented:
Light mode looks "unhighlighted" for some reason - probably need a different theme?
<img width="971" alt="Screenshot 2024-10-11 at 6 01 08 PM" src="https://github.com/user-attachments/assets/a9a6632c-9391-41c5-aad0-e43e26787466">

mikeldking reviewed (commented):
This is so cool. Excited to wire it all up

mikeldking commented on app/src/components/templateEditor/TemplateEditor.tsx:
```diff
@@ -0,0 +1,47 @@
+  }, [templateLanguage]);
+
+  return (
+    <CodeMirror theme={codeMirrorTheme} extensions={extensions} {...props} />
```
I see the ...props so I get that you are trying to keep this generic. Maybe you can expose under `/code` a version that doesn't have things like line numbers since I think that's the primary use-case

cephalization replied:
I'm happy to tweak any defaults you want for TemplateEditor, including removing line numbers and setting a default height.

I tend to opt for exposing all props on components so that I don't have to add them later but I'm happy with leaving extra props opaque until they're actually needed to be drilled through

cephalization replied:
<img width="922" alt="image" src="https://github.com/user-attachments/assets/28460f7d-2e09-4ec5-8eae-0d3a43286cfa">

No more gutter, line numbers, or line highlight by default

cephalization commented:
> Light mode looks "unhighlighted" for some reason - probably need a different theme?
> <img width="971" alt="Screenshot 2024-10-11 at 6 01 08 PM" src="https://github.com/user-attachments/assets/a9a6632c-9391-41c5-aad0-e43e26787466">
> 

Correct. Highlighting is determined by setting a semantic tag type and then the theme is meant to handle the actual coloring. We have no theme set on light mode, similar to the other CodeMirror editor. I'm happy to choose one but feel free to pick one as well and I'll add it

cephalization commented:
> I think there's some very minor syntax issues e.g. raw {{}}
> 
> <img width="457" alt="Screenshot 2024-10-11 at 5 59 36 PM" src="https://github.com/user-attachments/assets/2e7ad4bb-5f68-4ef2-8085-33247fb6f3b8">
> 

Mustache surprisingly has more parsing quirks than f-string. I'll spend a bit more time refining it but I'll have to be careful not to rabbit hole on it, I keep nerd sniping myself into optimization and tweaking on this

cephalization commented:
Alright I've fixed all of the parsing issues I could find for both langs, and implemented `\` escaping and fstring `{{escape}}` as well as mustache `{{{ escape }}}`.

The final parse "issue" that I will leave unresolved for now is `{test {value}}` The highlighter stops at the second to last brace, and the emitted variable name is `test {value`

cephalization commented on app/src/components/templateEditor/language/fStringTemplating/fStringTemplating.ts:
```diff
@@ -0,0 +1,63 @@
+import { LanguageSupport, LRLanguage } from "@codemirror/language";
```
Renamed

cephalization commented:
<img width="914" alt="image" src="https://github.com/user-attachments/assets/b5ff8fed-042a-4ca4-93fc-b8a0756bce79">

Light mode theme "github"
